### PR TITLE
James 1954 CommandInjectionDetectedException is raised by valid emails

### PR DESCRIPTION
**Pull request by chibenwa**
https://github.com/linagora/james-project/issues/668

We need context to not arbitrary reject valid email data
I also improved command injection detection mechanismes to allow RCPT like starttls@any.com

### CODE REVIEW

**Comment by chibenwa** [comment]
I need to : 

  -  improve detection algorithm to ensure we don't reject valid command that contains starttls (eg RCPT TO: starttls@domain.com )
  -  Disable it on DATA (should be able to process arbitrary datas + also speed ^^)


For this I need to retrieve SMTPSession that is the Attachment of the context of the core handler (sorry, I don't think we can do something else...).

I added unit / integration tests where needed, so that you can trust in what I've done.

**Comment by aduprat** [comment]
test this please

File: protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java [hunk 1]
```diff
@@ -189,6 +189,22 @@ public void startTlsShouldFailWhenFollowedByInjectedCommand() throws Exception {
         client.sendCommand("EHLO localhost");
 
         client.sendCommand("STARTTLS\r\nRSET\r\n");
+        System.out.println(client.getReplyString());
```
Comment (quynhn) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: Do we need this?

File: protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java [hunk 2]
```diff
@@ -189,6 +189,22 @@ public void startTlsShouldFailWhenFollowedByInjectedCommand() throws Exception {
...
+        ProtocolServer server = createServer(createProtocol(Optional.<ProtocolHandler> absent()), Encryption.createStartTls(BogusSslContextFactory.getServerContext()));
+        server.bind();
+
+        SMTPSClient client = createClient();
+        InetSocketAddress bindedAddress = new ProtocolServerUtils(server).retrieveBindedAddress();
+        client.connect(bindedAddress.getAddress().getHostAddress(), bindedAddress.getPort());
+        client.sendCommand("EHLO localhost");
+
+        client.sendCommand("RSET\r\nSTARTTLS\r\nRSET\r\n");
+        System.out.println(client.getReplyString());
```
Comment (quynhn) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: ?

Comment (quynhn) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: Do we need to call?
```
client.quit();
client.disconnect();
```

File: protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java (hunk 1 above)
Comment (chibenwa) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: Nope, it's debug... Sorry about this. Will remove it.

File: protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java (hunk 2 above)
Comment (chibenwa) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: Will add try/final statements here and everywhere else.

File: protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java [hunk 3]
```diff
@@ -98,48 +88,57 @@ private Protocol createProtocol(Optional<ProtocolHandler> handler) throws Wiring
...
-        client.disconnect();
+        try {
+            server.bind();
+            InetSocketAddress bindedAddress = new ProtocolServerUtils(server).retrieveBindedAddress();
+            client.connect(bindedAddress.getAddress().getHostAddress(), bindedAddress.getPort());
+            assertThat(SMTPReply.isPositiveCompletion(client.getReplyCode())).isTrue();
+        } finally {
+            client.quit();
+            client.disconnect();
+            server.unbind();
```
Comment (rouazana) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: why not putting this in @After?

Comment (chibenwa) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: Server is test depandant.
We could try to close after a null check in @after yes

Comment (rouazana) [inline protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyStartTlsSMTPServerTest.java]: great, this would avoid this boilerplate

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**